File: dotnet/src/xjustiz.core-dotnet/Util/Converter/XJustizConverter.cs
namespace xjustiz.core_dotnet.Util.Converter;

using System;
using System.Collections;
using System.Collections.Generic;
using System.Reflection;
using xjustiz.core_dotnet.Models;
using xjustiz.core_dotnet.Util.Versioning;

/// <summary>
/// Converter for X.Justiz and X.Justiz-Core messages.
/// Converts messages to a specific target version, filtering out unsupported properties.
/// </summary>
public static class XJustizConverter
{
    /// <summary>
    /// Converts the source message to the target X.Justiz version.
    /// </summary>
    /// <param name="source">The source message to convert.</param>
    /// <param name="targetVersion">The target X.Justiz version.</param>
    /// <returns>A result containing the converted message and a list of lost data.</returns>
    public static ConversionResult Convert(UebermittlungSchriftgutobjekteNachricht source, XJustizVersion targetVersion)
    {
        return ConvertInternal(source, targetVersion);
    }

    /// <summary>
    /// Converts the source message to the target X.Justiz-Core version.
    /// </summary>
    /// <param name="source">The source message to convert.</param>
    /// <param name="targetVersion">The target X.Justiz-Core version.</param>
    /// <returns>A result containing the converted message and a list of lost data.</returns>
    public static ConversionResult Convert(UebermittlungSchriftgutobjekteNachricht source, XJustizCoreVersion targetVersion)
    {
        return ConvertInternal(source, targetVersion);
    }

    /// <summary>
    /// Deep copies the object and filters properties based on version availability.
    /// </summary>
    /// <param name="source">The source.</param>
    /// <param name="targetVersion">The target version.</param>
    /// <param name="lostData">The lost data list.</param>
    /// <param name="context">The context path.</param>
    /// <returns>The copied object.</returns>
    internal static object? DeepCopy(object? source, Enum targetVersion, List<string> lostData, string context)
    {
        if (source == null)
        {
            return null;
        }

        var type = source.GetType();

        // 1. Check availability of the Type (Class/Enum) itself
        if (!IsSimple(type) && !IsAvailable(type, targetVersion))
        {
            lostData.Add($"Type '{type.Name}' is not available in {targetVersion.GetType().Name} {targetVersion}. (Context: {context})");
            return null;
        }

        // 2. Handle Simple Types (Primitives, Strings, Enums, DateTime, etc.)
        if (IsSimple(type))
        {
            return source;
        }

        // 3. Handle Collections (Arrays and Lists)
        if (typeof(IEnumerable).IsAssignableFrom(type))
        {
            return HandleCollection(source, type, targetVersion, lostData, context);
        }

        // 4. Handle Complex Objects (Classes)
        return HandleComplexObject(source, type, targetVersion, lostData, context);
    }

    internal static ConversionResult ConvertInternal(UebermittlungSchriftgutobjekteNachricht source, Enum targetVersion)
    {
        var lostData = new List<string>();
        // Root context name
        var result = (UebermittlungSchriftgutobjekteNachricht?)DeepCopy(source, targetVersion, lostData, typeof(UebermittlungSchriftgutobjekteNachricht).Name);

        return new ConversionResult
        {
            Result = result,
            LostData = lostData,
        };
    }

    private static object HandleCollection(object source, Type type, Enum targetVersion, List<string> lostData, string context)
    {
        if (type.IsArray)
        {
            var arr = (Array)source;
            var elementType = type.GetElementType() !;
            var validItems = new List<object?>();

            for (int i = 0; i < arr.Length; i++)
            {
                var val = arr.GetValue(i);
                var converted = DeepCopy(val, targetVersion, lostData, $"{context}[{i}]");
                if (converted != null)
                {
                    validItems.Add(converted);
                }
            }

            var newArr = Array.CreateInstance(elementType, validItems.Count);
            for (int i = 0; i < validItems.Count; i++)
            {
                newArr.SetValue(validItems[i], i);
            }

            return newArr;
        }
        else if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(List<>))
        {
            var list = (IList)Activator.CreateInstance(type) !;
            var sourceList = (IList)source;
            int i = 0;
            foreach (var item in sourceList)
            {
                var converted = DeepCopy(item, targetVersion, lostData, $"{context}[{i}]");
                if (converted != null)
                {
                    list.Add(converted);
                }

                i++;
            }

            return list;
        }

        return source;
    }

    private static object HandleComplexObject(object source, Type type, Enum targetVersion, List<string> lostData, string context)
    {
        var instance = Activator.CreateInstance(type);
        if (instance == null)
        {
            return null!;
        }

        foreach (var prop in type.GetProperties(BindingFlags.Public | BindingFlags.Instance))
        {
            if (!prop.CanRead || !prop.CanWrite || prop.GetIndexParameters().Length > 0)
            {
                continue;
            }

            // Check Property Availability
            if (!IsAvailable(prop, targetVersion))
            {
                var value = prop.GetValue(source);
                if (value != null)
                {
                    lostData.Add($"Property '{prop.Name}' in '{type.Name}' is not available in {targetVersion.GetType().Name} {targetVersion}. (Context: {context}.{prop.Name})");
                }

                continue;
            }

            var val = prop.GetValue(source);
            var convertedVal = DeepCopy(val, targetVersion, lostData, $"{context}.{prop.Name}");

            prop.SetValue(instance, convertedVal);
        }

        return instance;
    }

    private static bool IsSimple(Type type)
    {
        return type.IsPrimitive
            || type.IsEnum
            || type == typeof(string)
            || type == typeof(decimal)
            || type == typeof(DateTime)
            || type == typeof(Guid)
            || (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>) && IsSimple(Nullable.GetUnderlyingType(type) !));
    }

    private static bool IsAvailable(MemberInfo member, Enum targetVersion)
    {
        if (targetVersion is XJustizVersion xv)
        {
            var attr = member.GetCustomAttribute<XJustizAvailabilityAttribute>();
            if (attr != null)
            {
                return attr.IsAvailableIn(xv);
            }
        }
        else if (targetVersion is XJustizCoreVersion xcv)
        {
            var attr = member.GetCustomAttribute<XJustizCoreAvailabilityAttribute>();
            if (attr != null)
            {
                return attr.IsAvailableIn(xcv);
            }
        }

        return true;
    }
}

File: sdk/java/src/main/java/de/xjustiz/core/util/converter/XJustizConverter.java
package de.xjustiz.core.util.converter;

import de.xjustiz.core.models.UebermittlungSchriftgutobjekteNachricht;
import de.xjustiz.core.util.versioning.*;

import java.lang.reflect.*;
import java.time.temporal.Temporal;
import java.util.*;

/**
 * Konverter für Nachrichten im X.Justiz und X.Justiz Core Format.<br/>
 * Konvertiert Nachrichten in eine bestimmte Zielversion und filtert nicht
 * unterstützte Eigenschaften heraus.<br/>
 * <u><b>XJustiz Converter:</b></u> Converter for X.Justiz and X.Justiz Core
 * messages.
 * Converts messages to a specific target version, filtering out unsupported
 * properties.
 */
public final class XJustizConverter {

    private XJustizConverter() {
        // Utility class
    }

    /**
     * Konvertiert die Quellnachricht in die Ziel-X.Justiz-Version.<br/>
     * <u><b>Convert:</b></u> Converts the source message to the target X.Justiz
     * version.
     * 
     * @param source        The source message to convert.
     * @param targetVersion The target X.Justiz version.
     * @return A result containing the converted message and a list of lost data.
     */
    public static ConversionResult convert(UebermittlungSchriftgutobjekteNachricht source,
            XJustizVersion targetVersion) {
        return convertInternal(source, targetVersion);
    }

    /**
     * Konvertiert die Quellnachricht in die Ziel X.Justiz Core Version.<br/>
     * <u><b>Convert:</b></u> Converts the source message to the target X.Justiz
     * Core version.
     * 
     * @param source        The source message to convert.
     * @param targetVersion The target X.Justiz Core version.
     * @return A result containing the converted message and a list of lost data.
     */
    public static ConversionResult convert(UebermittlungSchriftgutobjekteNachricht source,
            XJustizCoreVersion targetVersion) {
        return convertInternal(source, targetVersion);
    }

    private static ConversionResult convertInternal(UebermittlungSchriftgutobjekteNachricht source,
            Enum<?> targetVersion) {
        List<String> lostData = new ArrayList<>();
        UebermittlungSchriftgutobjekteNachricht result = (UebermittlungSchriftgutobjekteNachricht) deepCopy(source,
                targetVersion, lostData, UebermittlungSchriftgutobjekteNachricht.class.getSimpleName());
        return new ConversionResult(result, lostData);
    }

    /**
     * Kopiert das Objekt und filtert Eigenschaften basierend auf der
     * Versionsverfügbarkeit.<br/>
     * <u><b>Deep copy:</b></u> Deep copies the object and filters properties based
     * on version availability.
     * 
     * @param source        The source object.
     * @param targetVersion The target version (XJustizVersion or
     *                      XJustizCoreVersion).
     * @param lostData      The lost data list to populate.
     * @param context       The context path for error messages.
     * @return The copied object, or null if not available.
     */
    public static Object deepCopy(Object source, Enum<?> targetVersion, List<String> lostData, String context) {
        if (source == null) {
            return null;
        }

        Class<?> type = source.getClass();

        // 1. Check availability of the Type (Class/Enum) itself
        if (!isSimple(type) && !isAvailable(type, targetVersion)) {
            lostData.add(String.format("Type '%s' is not available in %s %s. (Context: %s)",
                    type.getSimpleName(), targetVersion.getClass().getSimpleName(), targetVersion, context));
            return null;
        }

        // 2. Handle Simple Types (Primitives, Strings, Enums, DateTime, etc.)
        if (isSimple(type)) {
            return source;
        }

        // 3. Handle Collections (Arrays and Lists)
        if (type.isArray()) {
            return handleArray(source, targetVersion, lostData, context);
        }
        if (Collection.class.isAssignableFrom(type)) {
            return handleCollection(source, targetVersion, lostData, context);
        }

        // 4. Handle Complex Objects (Classes)
        return handleComplexObject(source, type, targetVersion, lostData, context);
    }

    private static Object handleArray(Object source, Enum<?> targetVersion, List<String> lostData, String context) {
        int length = Array.getLength(source);
        List<Object> validItems = new ArrayList<>();
        Class<?> componentType = source.getClass().getComponentType();

        for (int i = 0; i < length; i++) {
            Object val = Array.get(source, i);
            Object converted = deepCopy(val, targetVersion, lostData, context + "[" + i + "]");
            if (converted != null) {
                validItems.add(converted);
            }
        }

        Object newArray = Array.newInstance(componentType, validItems.size());
        for (int i = 0; i < validItems.size(); i++) {
            Array.set(newArray, i, validItems.get(i));
        }
        return newArray;
    }

    private static Object handleCollection(Object source, Enum<?> targetVersion, List<String> lostData,
            String context) {
        Collection<?> sourceCollection = (Collection<?>) source;
        Collection<Object> result;

        if (source instanceof List) {
            result = new ArrayList<>();
        } else if (source instanceof Set) {
            result = new LinkedHashSet<>();
        } else {
            result = new ArrayList<>();
        }

        int i = 0;
        for (Object item : sourceCollection) {
            Object converted = deepCopy(item, targetVersion, lostData, context + "[" + i + "]");
            if (converted != null) {
                result.add(converted);
            }
            i++;
        }
        return result;
    }

    private static Object handleComplexObject(Object source, Class<?> type, Enum<?> targetVersion,
            List<String> lostData, String context) {
        Object instance;
        try {
            Constructor<?> constructor = type.getDeclaredConstructor();
            constructor.setAccessible(true);
            instance = constructor.newInstance();
        } catch (Exception e) {
            return null;
        }

        List<Field> allFields = getAllFields(type);

        for (Field field : allFields) {
            if (Modifier.isStatic(field.getModifiers()) || Modifier.isFinal(field.getModifiers())) {
                continue;
            }

            field.setAccessible(true);

            // Check Field Availability
            if (!isAvailable(field, targetVersion)) {
                Object value = getFieldValue(field, source);
                if (value != null) {
                    lostData.add(String.format("Field '%s' in '%s' is not available in %s %s. (Context: %s.%s)",
                            field.getName(), type.getSimpleName(),
                            targetVersion.getClass().getSimpleName(), targetVersion,
                            context, field.getName()));
                }
                continue;
            }

            Object val = getFieldValue(field, source);
            Object convertedVal = deepCopy(val, targetVersion, lostData, context + "." + field.getName());

            try {
                field.set(instance, convertedVal);
            } catch (IllegalAccessException ignored) {
                // Continue if we can't set the field
            }
        }

        return instance;
    }

    private static List<Field> getAllFields(Class<?> type) {
        List<Field> fields = new ArrayList<>();
        Class<?> current = type;
        // Only traverse classes in our packages, not JDK classes
        while (current != null && current != Object.class && isXJustizModel(current)) {
            fields.addAll(Arrays.asList(current.getDeclaredFields()));
            current = current.getSuperclass();
        }
        return fields;
    }

    private static Object getFieldValue(Field field, Object obj) {
        try {
            // Skip synthetic fields
            if (field.isSynthetic()) {
                return null;
            }
            field.setAccessible(true);
            return field.get(obj);
        } catch (Exception e) {
            return null;
        }
    }

    private static boolean isXJustizModel(Class<?> type) {
        String packageName = type.getPackageName();
        return packageName != null && packageName.startsWith("de.xjustiz");
    }

    private static boolean isSimple(Class<?> type) {
        return type.isPrimitive()
                || type.isEnum()
                || type == String.class
                || Number.class.isAssignableFrom(type)
                || type == Boolean.class
                || type == Character.class
                || Temporal.class.isAssignableFrom(type)
                || type == Date.class
                || type == UUID.class
                || isWrapperType(type);
    }

    private static boolean isWrapperType(Class<?> type) {
        return type == Integer.class || type == Long.class || type == Double.class
                || type == Float.class || type == Short.class || type == Byte.class;
    }

    private static boolean isAvailable(AnnotatedElement member, Enum<?> targetVersion) {
        if (targetVersion instanceof XJustizVersion xv) {
            XJustizAvailability attr = member.getAnnotation(XJustizAvailability.class);
            if (attr != null) {
                return VersionAvailabilityChecker.isAvailable(
                        xv.getValue(),
                        attr.introduced().getValue(),
                        attr.removed() == XJustizVersion.NONE ? 0 : attr.removed().getValue());
            }
        } else if (targetVersion instanceof XJustizCoreVersion xcv) {
            XJustizCoreAvailability attr = member.getAnnotation(XJustizCoreAvailability.class);
            if (attr != null) {
                int removedValue = attr.isRemoved() ? attr.removed().getValue() : 0;
                return VersionAvailabilityChecker.isAvailable(
                        xcv.getValue(),
                        attr.introduced().getValue(),
                        removedValue);
            }
        }
        return true;
    }
}
